plugins/arrow-flight-rpc/src/main/java/org/opensearch/arrow/flight/impl/BaseFlightProducer.java

```diff
@@ -0,0 +1,283 @@
+            throw CallStatus.INVALID_ARGUMENT.withDescription("Invalid descriptor format: " + e.getMessage())
+                .withCause(e)
+                .toRuntimeException();
+        }
+
+        if (streamTicket.getNodeId().equals(flightClientManager.getLocalNodeId())) {
+            return getLocalFlightInfo(streamTicket, descriptor);
+        } else {
+            return getRemoteFlightInfo(streamTicket, descriptor);
+        }
+    }
+
+    private FlightStreamManager.StreamProducerHolder acquireStreamProducer(
+        StreamTicket streamTicket,
+        Ticket ticket,
+        ServerStreamListener listener
+    ) {
+        if (streamTicket.getNodeId().equals(flightClientManager.getLocalNodeId())) {
+            return streamManager.removeStreamProducer(streamTicket).orElse(null);
+        }
+        Optional<FlightClient> remoteClient = flightClientManager.getFlightClient(streamTicket.getNodeId());
+        if (remoteClient.isPresent()) {
+            try {
+                StreamProducer<VectorSchemaRoot, BufferAllocator> proxyProvider = createProxyProducer(remoteClient.get(), ticket, listener);
+                if (proxyProvider != null) {
+                    return FlightStreamManager.StreamProducerHolder.create(proxyProvider, allocator);
+                }
+            } catch (Exception e) {
+                logger.error("Failed to create stream producer", e);
+                listener.error(
+                    CallStatus.INTERNAL.withDescription("Failed to create stream producer: " + e.getMessage())
+                        .withCause(e)
+                        .toRuntimeException()
+                );
+            }
+        }
+        return null;
+    }
+
+    private StreamProducer<VectorSchemaRoot, BufferAllocator> createProxyProducer(
+        FlightClient remoteClient,
+        Ticket ticket,
+        ServerStreamListener listener
+    ) {
+
+        try (FlightStream flightStream = remoteClient.getStream(ticket)) {
+            if (flightStream == null) {
+                logger.error("Failed to obtain flight stream");
+                listener.error(CallStatus.INTERNAL.withDescription("Failed to create remote flight stream").toRuntimeException());
+                return null;
+            }
+            return new ProxyStreamProducer(new FlightStreamReader(flightStream));
+        } catch (Exception e) {
+            logger.error("Error creating proxy producer", e);
+            listener.error(
+                CallStatus.INTERNAL.withDescription("Failed to create proxy producer: " + e.getMessage()).withCause(e).toRuntimeException()
+            );
+            return null;
+        }
+    }
+
+    private void processStream(
+        FlightStreamManager.StreamProducerHolder streamProducerHolder,
+        StreamProducer.BatchedJob<VectorSchemaRoot> batchedJob,
+        CallContext context,
+        ServerStreamListener listener
+    ) {
+        listener.setOnCancelHandler(batchedJob::onCancel);
```

Comment:
I don't think you need, `backpressureStrategy.register` will call `listener:setOnCancelHandler`

Reply:
nice! removed